File: leetcode_problems/typescript/src/lc1774_closest_desert_price.ts
function closestCost(
  baseCosts: number[],
  toppingCosts: number[],
  target: number
): number {
  function get_min(total: number) {
    let sp1 = target - total;
    let sp2 = target - min;
    if (Math.abs(sp1) > Math.abs(sp2)) {
      return min;
    } else if (Math.abs(sp1) < Math.abs(sp2)) {
      return total;
    } else {
      return sp1 > sp2 ? total : min;
    }
  }

  function backtrack(i: number) {
    if (total >= target) {
      min = get_min(total);
      return;
    }
    if (i >= toppingCosts.length) {
      min = get_min(total);
      return;
    }

    backtrack(i + 1);
    total += toppingCosts[i];
    backtrack(i + 1);
    total += toppingCosts[i];
    backtrack(i + 1);
    total -= 2 * toppingCosts[i];
  }

  let min = 10 ** 6;
  let total = 0;
  for (let bc of baseCosts) {
    total += bc;
    backtrack(0);
    total -= bc;
  }
  return min;
}

File: leetcode_problems/cpp/lc1774_closest_desert_price.cpp
#include <vector>
#include <algorithm>
using std::vector;

class Solution
{
public:
    int get_min(int total, int target, int min)
    {
        int sp1 = target - total;
        int sp2 = target - min;
        if (std::abs(sp1) > std::abs(sp2))
        {
            return min;
        }
        else if (std::abs(sp1) < std::abs(sp2))
        {
            return total;
        }
        else
        {
            return sp1 > sp2 ? total : min;
        }
    }

    void bactrack(int i, int &total, int &target, int &min, vector<int> &tc)
    {
        if (total >= target)
        {
            min = get_min(total, target, min);
            return;
        }
        if (i >= tc.size())
        {
            min = get_min(total, target, min);
            return;
        }

        bactrack(i + 1, total, target, min, tc);
        total += tc[i];
        bactrack(i + 1, total, target, min, tc);
        total += tc[i];
        bactrack(i + 1, total, target, min, tc);
        total -= 2 * tc[i];
    }

    int closestCost(vector<int> &baseCosts, vector<int> &toppingCosts, int target)
    {
        int min = 100000000;
        int total = 0;
        for (int bc : baseCosts)
        {
            total += bc;
            bactrack(0, total, target, min, toppingCosts);
            total -= bc;
        }
        return min;
    }
};